milli/src/update/index_documents/mod.rs

```diff
@@ -55,120 +51,115 @@ pub enum IndexDocumentsMethod {
+}
+
+#[derive(Default, Debug, Clone)]
+pub struct IndexDocumentsConfig {
+    pub facet_level_group_size: Option<NonZeroUsize>,
+    pub facet_min_level_size: Option<NonZeroUsize>,
+    pub words_prefix_threshold: Option<u32>,
+    pub max_prefix_length: Option<usize>,
+    pub words_positions_level_group_size: Option<NonZeroU32>,
+    pub words_positions_min_level_size: Option<NonZeroU32>,
+    pub update_method: IndexDocumentsMethod,
+    pub autogenerate_docids: bool,
 }
 
-impl<'t, 'u, 'i, 'a> IndexDocuments<'t, 'u, 'i, 'a> {
+impl<'t, 'u, 'i, 'a, F> IndexDocuments<'t, 'u, 'i, 'a, F>
+where
+    F: Fn(UpdateIndexingStep) + Sync,
+{
     pub fn new(
         wtxn: &'t mut heed::RwTxn<'i, 'u>,
         index: &'i Index,
-    ) -> IndexDocuments<'t, 'u, 'i, 'a> {
+        indexer_config: &'a IndexerConfig,
+        config: IndexDocumentsConfig,
+        progress: F,
+    ) -> IndexDocuments<'t, 'u, 'i, 'a, F> {
+        let transform = Some(Transform::new(
+            &index,
+            indexer_config,
+            config.update_method,
+            config.autogenerate_docids,
+        ));
+
         IndexDocuments {
+            transform,
+            config,
+            indexer_config,
+            progress,
             wtxn,
             index,
-            log_every_n: None,
-            documents_chunk_size: None,
-            max_nb_chunks: None,
-            max_memory: None,
-            chunk_compression_type: CompressionType::None,
-            chunk_compression_level: None,
-            thread_pool: None,
-            facet_level_group_size: None,
-            facet_min_level_size: None,
-            words_prefix_threshold: None,
-            max_prefix_length: None,
-            words_positions_level_group_size: None,
-            words_positions_min_level_size: None,
-            update_method: IndexDocumentsMethod::ReplaceDocuments,
-            autogenerate_docids: false,
-            max_positions_per_attributes: None,
+            added_documents: 0,
         }
     }
 
-    pub fn log_every_n(&mut self, n: usize) {
-        self.log_every_n = Some(n);
-    }
-
-    pub fn index_documents_method(&mut self, method: IndexDocumentsMethod) {
-        self.update_method = method;
-    }
-
-    pub fn enable_autogenerate_docids(&mut self) {
-        self.autogenerate_docids = true;
-    }
-
-    pub fn disable_autogenerate_docids(&mut self) {
-        self.autogenerate_docids = false;
-    }
-
-    #[logging_timer::time("IndexDocuments::{}")]
-    pub fn execute<R, F>(
-        self,
-        reader: DocumentBatchReader<R>,
-        progress_callback: F,
-    ) -> Result<DocumentAdditionResult>
+    /// Adds a batch of documents to the current builder.
+    ///
+    /// Since the documents are progressively added to the writer, a failure will cause a stale
+    /// builder, and the builder must be discarded.
+    #[must_use = "a builder must not be used after a add document operation reported an error."]
```

Comment:
It is not how `must_use` works, [here it is advertising the user that he is not using the u64 output by this method](https://rust-lang.github.io/rfcs/1940-must-use-functions.html).

Reply:
Can you add a point about what this return `u64` is, please?